### DISCUSSION TITLE
Reduce resolution of image in 2d array

### DISCUSSION THREAD

**Discussion by Isaac-Leonard** [rust-ndarray/ndarray#1325] [Q&A] [answered]
I've got pixel data in a 2 dimensional array and I want to reduce the resolution by splitting into chunks and running interpolation on it such as using the .mean() method.
So something along the lines of
```
let img = img.chunk_2d(x_scale, y_scale).map(|chunk|c hunk.mean());
```
The chunk_2d method doesn't exist of course, is there a similar method that would do what I want or how would I go about doing this?
I've looked at the .into_shap() method but it doesn't seem to do what I want exactly.

**Comment by Isaac-Leonard** [comment]
Also the .chunks_exact() method doesn't seem to give a good way to resolve back into a 2d array and only gives one dimension to chunk the data by so can't have different x and y scales

**Comment by nilgoyette** [comment]
Would this work?

    let mean = Zip::from(img.exact_chunks((x_scale, y_scale))).map_collect(|chunk| chunk.mean());

[exact_chunks](https://docs.rs/ndarray/latest/ndarray/struct.ArrayBase.html#method.exact_chunks) is a producer, so it can be used with [Zip](https://docs.rs/ndarray/latest/ndarray/struct.Zip.html). We usually zip over more than one array, but this is not mandatory. `Zip` is a wonderful tool! I suggest you to read it's doc.

**Comment by Isaac-Leonard** [comment]
It looks like that works, thank you
Only tangentially related but would you know how I could make this so I'm not iterating over references to the pixels in:
```
    for row in data.rows() {
        for (i, pixel) in row.into_iter().enumerate() {}
    }
```
I'd rather iterate over the raw values Or do you think dereferencing wouldn't matter performance wise?

**Reply by nilgoyette** [reply, marked as answer]
I can't remember where this is written, but you should use iter/into_iter as less as possible because it is slow. For example, what you wrote can be expressed as

    Zip::indexed(&data).for_each(|idx, &pixel| {
        if idx == (0, 0) { wathever }
    })

If you're starting with ndarray, I strongly suggest you to read the [quickstart](https://github.com/rust-ndarray/ndarray/blob/master/README-quick-start.md) and [official doc](https://docs.rs/ndarray/latest/ndarray/struct.ArrayBase.html).

**Comment by Isaac-Leonard** [comment]
Awesome, thank you so much and yeah I'll do that now.
I kinda just jumped straight in lol
